### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by diogocaribe** [cbalint13/gdal-segment#12]
Why my segmentation show me this error and not finish the process?

```
~/gdal-segment/bin $ gdal-segment -algo SLIC -region 500 -niter 2 /tmp/tmp6er1q9dg/cut.tif -out output-slic.shp
Segments raster using: SLIC
Process use parameter: region=500 niter=2

Load Raster #1 (#1): /tmp/tmp6er1q9dg/cut.tif
  Scene have [3] bands
  CH: #001 chann: (#1 Band | [Int16])
           areas: (930 Pixels x 843 Lines) pixels
           tiles: (1 Columns x 843 Rows) blocks
           block: (930 Pixels x 1 Lines) pixels / tile
           0...10...20...30...40...50...60...70...80...90...100 - done.
  CH: #002 chann: (#2 Band | [Int16])
           areas: (930 Pixels x 843 Lines) pixels
           tiles: (1 Columns x 843 Rows) blocks
           block: (930 Pixels x 1 Lines) pixels / tile
           0...10...20...30...40...50...60...70...80...90...100 - done.
  CH: #003 chann: (#3 Band | [Int16])
           areas: (930 Pixels x 843 Lines) pixels
           tiles: (1 Columns x 843 Rows) blocks
           block: (930 Pixels x 1 Lines) pixels / tile
           0...10...20...30...40...50...60...70...80...90...100 - done.
Time: 0.064325 sec

Init Superpixels
Time: 0.080034 sec

Grow Superpixels: #2 iterations
           inits: 4 superpixels
           count: 4 superpixels (growed in 0.436875 sec)
           final: 1 superpixels (merged in 0.094649 sec)
Time: 0.531673 sec

Parse edges in segmented image
0...10...20...30...40...50...60...70...80...90...100 - done.
Time: 0.018492 sec

Compute Statistics
       Computing CLASS polygons intensity
       0...10...20...30...40...50...60...70...80...90...100 - done.
       Computing CLASS averaged intensity
              Computing CLASS standard deviation
       0...10...20...30...40...50...60...70...80...90...100 - done.
       Normalize CLASS standard deviation
       Time: 13.460847 sec

Segmentation fault (core dumped)
```

**Comment by rodolfolotte** [comment]
Diogo, First, test it with different images and see what happen. Second, open the project and find the cpp file "raster.cpp" and look at ComputStats routine. Probably, is something going wrong here:

```
  printf ("       Normalize CLASS standard deviation\n");
  printf ("       ");

  for (int k = 0; k < labelpixels.rows; k++)
  {
      #pragma omp parallel for schedule(dynamic)
      for (int b = 0; b < m_bands; b++)
      {
          stdCH.at<double>(b,k) = sqrt(stdCH.at<double>(b,k) / labelpixels.at<u_int32_t>(k));
      }
      GDALTermProgress( (float)(k+1) / (float)(labelpixels.rows), NULL, NULL );
  }
  GDALTermProgress( 1.0f, NULL, NULL );
```

**Comment by cbalint13** [comment]
```
Grow Superpixels: #2 iterations
           inits: 4 superpixels
           count: 4 superpixels (growed in 0.436875 sec)
           final: 1 superpixels (merged in 0.094649 sec)
Time: 0.531673 sec
```

  - You choose wrong superpixel region size ```-region 500```. Region size is expressed in pixels, a region size of 8 means one superpixel will try cover a 8x8 pixel area. Your raster is 930x843 so a region size of 500 will divide all in max 4 regions, but seems that at final refine step all collapsed into single polygon (superpixel). Also 2 iteration is quite low, i would recommend minimum 10 iterations (steps that refine edges of superpixels).

**Comment by diogocaribe** [comment]
@cbalint13 and @rodolfolotte,

Thank you my friends for their help,

As a biologist, I am working so hard to understanding how to put this code to work well. But I think that is for a good reason - forest monitoring of my state.

When I post this problem, sample was showing the same problem. So, I reinstalled everything and sample worked well now. 

@cbalint13  Thank you for your explanation about parameters. I am using them intuitively. I changed them but the problem persisted. I will see what the problem with my raster. 

```
~/gdal-segment/samples $ ./1-segment.sh 
Segments raster using: SLIC
Process use parameter: region=10 niter=10

Load Raster #1 (#1): 25cm_orto0.jpg
  Scene have [3] bands
  CH: #001 chann: (#1 Band | [Byte])
           areas: (700 Pixels x 429 Lines) pixels
           tiles: (1 Columns x 429 Rows) blocks
           block: (700 Pixels x 1 Lines) pixels / tile
           0...10...20...30...40...50...60...70...80...90...100 - done.
  CH: #002 chann: (#2 Band | [Byte])
           areas: (700 Pixels x 429 Lines) pixels
           tiles: (1 Columns x 429 Rows) blocks
           block: (700 Pixels x 1 Lines) pixels / tile
           0...10...20...30...40...50...60...70...80...90...100 - done.
  CH: #003 chann: (#3 Band | [Byte])
           areas: (700 Pixels x 429 Lines) pixels
           tiles: (1 Columns x 429 Rows) blocks
           block: (700 Pixels x 1 Lines) pixels / tile
           0...10...20...30...40...50...60...70...80...90...100 - done.
Time: 0.059975 sec

Convert to LAB colorspace.
Init Superpixels
Time: 0.033653 sec

Grow Superpixels: #10 iterations
           inits: 3010 superpixels
           count: 3010 superpixels (growed in 2.266113 sec)
           final: 2896 superpixels (merged in 0.041434 sec)
Time: 2.307666 sec

Parse edges in segmented image
0...10...20...30...40...50...60...70...80...90...100 - done.
Time: 0.030255 sec

Compute Statistics
       Computing CLASS polygons intensity
       0...10...20...30...40...50...60...70...80...90...100 - done.
       Computing CLASS averaged intensity
       0...10...20...30...40...50...60...70...80...90...100 - done.
       Computing CLASS standard deviation
       0...10...20...30...40...50...60...70...80...90...100 - done.
       Normalize CLASS standard deviation
       0...10...20...30...40...50...60...70...80...90...100 - done.
Time: 5.129208 sec

Write File: output-slic.shp (polygon)

```

**Comment by diogocaribe** [comment]
My friends,

One things that is that my landsat was in UInt16. I have changed it to Byte and the segmentation works well.

:+1: 

Regards!
